### PR TITLE
Yamaha: Add test fragment. Tests handlers with xml protocol

### PR DESCRIPTION
**Pull request by davidgraeff**
Necessary before the json protocol can be added.

Signed-off-by: David Gräff <david.graeff@web.de>

### CODE REVIEW

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/BridgeHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    public void setUp() throws InterruptedException {
+        initMocks(this);
+
+        emulatedReceiver = new EmulatedYamahaReceiver(12312);
+        assertTrue(emulatedReceiver.waitForStarted(1000));
+
+        ThingUID thingUID = YamahaDiscoveryParticipant.getThingUID(YamahaReceiverBindingConstants.UPNP_MANUFACTURER,
+                YamahaReceiverBindingConstants.UPNP_TYPE, "test");
+        when(thing.getUID()).thenReturn(thingUID);
+        // when(thing.getBridgeUID()).thenReturn(thingUID);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/BridgeHandlerTest.java]: Can this be removed?

File: addons/binding/org.openhab.binding.yamahareceiver.test/pom.xml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  </parent>
+
+  <groupId>org.eclipse.smarthome.archetype</groupId>
+  <artifactId>org.openhab.binding.yamahareceiver.test</artifactId>
+  <version>2.2.0-SNAPSHOT</version>
+  <packaging>eclipse-test-plugin</packaging>
+
+  <name>yamahareceiver.test Binding Tests</name>
+
+  <properties>
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/pom.xml]: is this from the artifact? It should not be needed.

File: addons/binding/org.openhab.binding.yamahareceiver.test/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Tests for the Yamaha Binding
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/README.md]: There should be a new line, not that this ends up in the readme but for completion it should be great.

File: addons/binding/org.openhab.binding.yamahareceiver.test/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ org.eclipse.smarthome.test.storage,
+ org.hamcrest;core=split,
+ org.junit;version="4.0.0",
+ org.mockito,
+ org.openhab.binding.yamahareceiver,
+ org.openhab.binding.yamahareceiver.handler,
+ org.osgi.framework,
+ org.osgi.service.device,
+ org.slf4j
+Require-Bundle: org.junit,
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/META-INF/MANIFEST.MF]: ;-)

File: addons/binding/org.openhab.binding.yamahareceiver.test/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,32 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name:  yamahareceiver.test Binding Tests
+Bundle-SymbolicName: org.openhab.binding.yamahareceiver.test;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.1.0.qualifier
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/META-INF/MANIFEST.MF]: Wrong bundle version

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                return "<Basic_Status><Power_Control><Power>On</Power></Power_Control><Input><Input_Sel>Net Radio</Input_Sel><Input_Sel_Item_Info><Src_Name>NET_RADIO</Src_Name></Input_Sel_Item_Info></Input><Surround><Program_Sel><Current><Sound_Program>7ch Stereo</Sound_Program></Current></Program_Sel></Surround><Volume><Mute>Off</Mute><Lvl><Val>100</Val></Lvl></Volume></Basic_Status>";
+            case "Input":
+                firstChild = firstChild.getFirstChild();
+                if (firstChild.getNodeName().equals("Input_Sel_Item")) {
+                    child = firstChild.getFirstChild();
+                    if (!child.getTextContent().equals("GetParam")) {
+                        throw new ParseException("Node for Input/Input_Sel_Item/ only supports GetParam", 0);
+                    }
+                    return "<Input><Input_Sel_Item><Item1><Param>HDMI 1</Param><RW>RW</RW></Item1><Item2><Param>Net Radio</Param><RW>RW</RW></Item2></Input_Sel_Item></Input>";
+                }
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java]: missing break if the if does not match

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+/**
+ * Emulates a Yamaha Receiver XML API endpoint (Http/XML on localhost:12121/YamahaRemoteControl/ctrl)
+ *
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java]: @author

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/TestDiscoveryService.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.model.types.DeviceType;
+import org.jupnp.model.types.UDN;
+import org.openhab.binding.yamahareceiver.YamahaReceiverBindingConstants;
+import org.openhab.binding.yamahareceiver.discovery.YamahaDiscoveryParticipant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TestDiscoveryService extends AbstractDiscoveryService {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/TestDiscoveryService.java]: Class-level docblock

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandlerExt.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    protected void updateAsyncMakeOfflineIfFail(IStateUpdateable stateUpdateable) {
+        try {
+            stateUpdateable.update();
+        } catch (IOException | ReceivedMessageParseException e) {
+            if (updateException == null) {
+                updateException = e;
+            }
+        }
+    }
+}
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandlerExt.java]: Missing newline

**Comment by martinvw** [comment]
Note that both travis & jenkins do not agree :-)

**Comment by martinvw** [comment]
```
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: org.openhab.binding.yamahareceiver.test 2.1.0.qualifier
[ERROR]   Missing requirement: org.openhab.binding.yamahareceiver.test 2.1.0.qualifier requires 'package org.eclipse.jdt.annotation 0.0.0' but it could not be found
[ERROR] 
```

**Comment by davidgraeff** [comment]
Travis and jenkins will not like this PR until the dependency PR is merged

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java (hunk 7 above)
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/EmulatedYamahaReceiver.java]: done

File: addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandlerExt.java (hunk 9 above)
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.yamahareceiver.test/src/test/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandlerExt.java]: done

**Comment by martinvw** [comment]
@davidgraeff since the other pr is merged can you rebase if needed?

**Comment by martinvw** [comment]
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:1.0.0:package-plugin (default-package-plugin) on project org.openhab.binding.yamahareceiver.test: /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.yamahareceiver.test/build.properties: bin.includes value(s) [OSGI-INF/] do not match any files. -> [Help 1]
```

I assume it's caused by the directory missing so maybe add a `.gitignore` in that directory or the workaround with the `.keep` file

**Comment by martinvw** [comment]
@davidgraeff what is the status of this PR?

**Comment by davidgraeff** [comment]
It is on hold until the ESH PRs are through

**Comment by martinvw** [comment]
> It is on hold until the ESH PRs are through

Thanks for the status update 👍

**Comment by wborn** [comment]
> It is on hold until the ESH PRs are through

Which ESH PRs is/was this PR waiting on @davidgraeff?

**Comment by davidgraeff** [comment]
Outdated

**Pull request closed without merging**